// @simplyjs/lib/dom/els.js
//various elements creation utility

import { create } from './base.js';

export var a = (opts, children) => create('a', opts, children);
export var abbr = (opts, children) => create('abbr', opts, children);
export var address = (opts, children) => create('address', opts, children);
export var area = (opts, children) => create('area', opts, children);
export var article = (opts, children) => create('article', opts, children);
export var aside = (opts, children) => create('aside', opts, children);
export var audio = (opts, children) => create('audio', opts, children);

export var b = (opts, children) => create('b', opts, children);
export var bdi = (opts, children) => create('bdi', opts, children);
export var bdo = (opts, children) => create('bdo', opts, children);
export var blockquote = (opts, children) => create('blockquote', opts, children);
export var body = (opts, children) => create('body', opts, children);
export var br = (opts, children) => create('br', opts, children);
export var button = (opts, children) => create('button', opts, children);

export var canvas = (opts, children) => create('canvas', opts, children);
export var caption = (opts, children) => create('caption', opts, children);
export var cite = (opts, children) => create('cite', opts, children);
export var code = (opts, children) => create('code', opts, children);

export var data = (opts, children) => create('data', opts, children);
export var datalist = (opts, children) => create('datalist', opts, children);
export var dd = (opts, children) => create('dd', opts, children);
export var del = (opts, children) => create('del', opts, children);
export var details = (opts, children) => create('details', opts, children);
export var dfn = (opts, children) => create('dfn', opts, children);
export var dialog = (opts, children) => create('dialog', opts, children);
export var div = (opts, children) => create('div', opts, children);
export var dl = (opts, children) => create('dl', opts, children);
export var dt = (opts, children) => create('dt', opts, children);

export var em = (opts, children) => create('em', opts, children);
export var embed = (opts, children) => create('embed', opts, children);

export var fieldset = (opts, children) => create('fieldset', opts, children);
export var figcaption = (opts, children) => create('figcaption', opts, children);
export var figure = (opts, children) => create('figure', opts, children);
export var footer = (opts, children) => create('footer', opts, children);
export var form = (opts, children) => create('form', opts, children);

export var h1 = (opts, children) => create('h1', opts, children);
export var h2 = (opts, children) => create('h2', opts, children);
export var h3 = (opts, children) => create('h3', opts, children);
export var h4 = (opts, children) => create('h4', opts, children);
export var h5 = (opts, children) => create('h5', opts, children);
export var h6 = (opts, children) => create('h6', opts, children);
export var header = (opts, children) => create('header', opts, children);
export var hgroup = (opts, children) => create('hgroup', opts, children);
export var hr = (opts, children) => create('hr', opts, children);
export var html = (opts, children) => create('html', opts, children);

export var i = (opts, children) => create('i', opts, children);
export var iframe = (opts, children) => create('iframe', opts, children);
export var img = (opts, children) => create('img', opts, children);
export var input = (opts, children) => create('input', opts, children);
export var ins = (opts, children) => create('ins', opts, children);

export var kbd = (opts, children) => create('kbd', opts, children);

export var label = (opts, children) => create('label', opts, children);
export var legend = (opts, children) => create('legend', opts, children);
export var li = (opts, children) => create('li', opts, children);

export var main = (opts, children) => create('main', opts, children);
export var map = (opts, children) => create('map', opts, children);
export var mark = (opts, children) => create('mark', opts, children);
export var menu = (opts, children) => create('menu', opts, children);
export var meta = (opts, children) => create('meta', opts, children);
export var meter = (opts, children) => create('meter', opts, children);

export var nav = (opts, children) => create('nav', opts, children);

export var ol = (opts, children) => create('ol', opts, children);
export var optgroup = (opts, children) => create('optgroup', opts, children);
export var option = (opts, children) => create('option', opts, children);
export var output = (opts, children) => create('output', opts, children);

export var p = (opts, children) => create('p', opts, children);
export var param = (opts, children) => create('param', opts, children);
export var picture = (opts, children) => create('picture', opts, children);
export var pre = (opts, children) => create('pre', opts, children);
export var progress = (opts, children) => create('progress', opts, children);

export var q = (opts, children) => create('q', opts, children);

export var rb = (opts, children) => create('rb', opts, children);
export var rp = (opts, children) => create('rp', opts, children);
export var rt = (opts, children) => create('rt', opts, children);
export var rtc = (opts, children) => create('rtc', opts, children);
export var ruby = (opts, children) => create('ruby', opts, children);

export var s = (opts, children) => create('s', opts, children);
export var samp = (opts, children) => create('samp', opts, children);
export var section = (opts, children) => create('section', opts, children);
export var select = (opts, children) => create('select', opts, children);
export var slot = (opts, children) => create('slot', opts, children);
export var small = (opts, children) => create('small', opts, children);
export var source = (opts, children) => create('source', opts, children);
export var span = (opts, children) => create('span', opts, children);
export var strong = (opts, children) => create('strong', opts, children);
export var style = (opts, children) => create('style', opts, children);
export var sub = (opts, children) => create('sub', opts, children);
export var summary = (opts, children) => create('summary', opts, children);
export var sup = (opts, children) => create('sup', opts, children);

export var table = (opts, children) => create('table', opts, children);
export var tbody = (opts, children) => create('tbody', opts, children);
export var td = (opts, children) => create('td', opts, children);
export var template = (opts, children) => create('template', opts, children);
export var textarea = (opts, children) => create('textarea', opts, children);
export var tfoot = (opts, children) => create('tfoot', opts, children);
export var th = (opts, children) => create('th', opts, children);
export var thead = (opts, children) => create('thead', opts, children);
export var time = (opts, children) => create('time', opts, children);
export var title = (opts, children) => create('title', opts, children);
export var tr = (opts, children) => create('tr', opts, children);
export var track = (opts, children) => create('track', opts, children);

export var u = (opts, children) => create('u', opts, children);
export var ul = (opts, children) => create('ul', opts, children);

export var Var = (opts, children) => create('var', opts, children);
export var video = (opts, children) => create('video', opts, children);

export var wbr = (opts, children) => create('wbr', opts, children);
